### PR TITLE
[rllib] Basic IMPALA implementation (using deepmind's reference vtrace.py)

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

- [x] Rename AsyncSamplesOptimizer -> AsyncReplayOptimizer
- [x] Add AsyncSamplesOptimizer that implements the IMPALA architecture
- [x] integrate V-trace with a3c policy graph
- [x] audit V-trace integration
- [x] benchmark compare vs A3C and with V-trace on/off

PongNoFrameskip-v4 on IMPALA scaling from 16 to 128 workers, solving Pong in <10 min. For reference, solving this env takes ~40 minutes for Ape-X and several hours for A3C.

impala

cc @joneswong 

## Related issue number

https://github.com/ray-project/ray/issues/1924

### CODE REVIEW

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+"""This is an variant of A3CPolicyGraph that uses V-trace for loss calc.
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: Adapted from a3c_policy_graph.py

File: python/ray/rllib/agents/impala/vtrace.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
+# Copyright 2018 Google LLC
```
Comment (ericl) [inline python/ray/rllib/agents/impala/vtrace.py]: Code inlined verbatim

File: python/ray/rllib/optimizers/async_replay_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
+"""Implements Distributed Prioritized Experience Replay.
```
Comment (ericl) [inline python/ray/rllib/optimizers/async_replay_optimizer.py]: Moved from async_sample_optimizer

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py (hunk 1 above)
Comment (robertnishihara) [reply python/ray/rllib/agents/impala/vtrace_policy_graph.py]: "an variant" -> "a variant"

File: python/ray/rllib/optimizers/async_samples_optimizer.py [hunk 4]
```diff
@@ -1,108 +1,28 @@
-"""Implements Distributed Prioritized Experience Replay.
+"""Implements the IMPALA architecture.
```
Comment (ericl) [inline python/ray/rllib/optimizers/async_samples_optimizer.py]: Adapted from https://github.com/ray-project/ray/pull/2147

**Comment by richardliaw** [comment]
Is the 16 workers A3C? Did you try A3C with vectorized envs too?

**Comment by ericl** [comment]
No, this is all impala. A3C takes hours to get to the same point.

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                discounts=tf.to_float(~dones) * discount,
+                rewards=rewards,
+                values=values,
+                bootstrap_value=bootstrap_value,
+                clip_rho_threshold=tf.cast(clip_rho_threshold, tf.float32),
+                clip_pg_rho_threshold=tf.cast(clip_pg_rho_threshold,
+                                              tf.float32))
+
+        # The policy gradients loss
+        self.pi_loss = -tf.reduce_sum(
```
Comment (richardliaw) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: these should be `reduce_mean` now that concatenation is removed.

**Review by richardliaw — APPROVED** [review]

File: doc/source/rllib-training.rst [hunk 6]
```diff
@@ -74,13 +74,19 @@ located at ``~/ray_results/default/DQN_CartPole-v0_0upjmdgr0/checkpoint-1``
...
 Some good hyperparameters and settings are available in
 `the repository <https://github.com/ray-project/ray/blob/master/python/ray/rllib/tuned_examples>`__
 (some of them are tuned to run on GPUs). If you find better settings or tune
 an algorithm on a different domain, consider submitting a Pull Request!
 
+You can run these with the ``train.py`` script as follows:
+
+.. code-block:: bash
+
+    python ray/python/ray/rllib/train.py -f /path/to/tuned/example.yaml
```
Comment (richardliaw) [inline doc/source/rllib-training.rst]: `rllib train -f ...`

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py (hunk 5 above)
Comment (ericl) [reply python/ray/rllib/agents/impala/vtrace_policy_graph.py]: The deepmind impl seems to use reduce_sum; we can either keep it or change it together with a3c.

File: doc/source/rllib-training.rst (hunk 6 above)
Comment (ericl) [reply doc/source/rllib-training.rst]: I don't think this works if you don't pip install, so going to keep it for now

**Comment by ericl** [comment]
Everything else looks good, merging after yapf.

**Pull request merged**